File: C++/version_1.0.0.cpp
#include <iostream>
#include <string>
#include <chrono>
#include <thread>
#include <httplib.h>
#include <nlohmann/json.hpp>

using json = nlohmann::json;

const std::string BOT_TOKEN = "paste token"; // Replace with your token
const std::string GUILD_ID = "paste server id"; // Replace with your server ID
const std::string MODMAIL_CATEGORY_NAME = "ModMail"; // create category named ModMail or change it urself
const std::string MODMAIL_LOG_CHANNEL = "modmail_log_channel_id"; // Replace with modmail-log channel ID

std::map<std::string, std::string> ticket_start_times;

std::string now() {
    auto t = std::time(nullptr);
    auto tm = *std::localtime(&t);
    char buffer[20];
    std::strftime(buffer, sizeof(buffer), "%Y-%m-%d %H:%M:%S", &tm);
    return std::string(buffer);
}

size_t WriteCallback(void* contents, size_t size, size_t nmemb, std::string* userp) {
    size_t total_size = size * nmemb;
    userp->append((char*)contents, total_size);
    return total_size;
}

void send_message(httplib::Client& client, const std::string& channel_id, const json& payload) {
    auto res = client.Post(("/channels/" + channel_id + "/messages").c_str(),
                           "Bot " + BOT_TOKEN,
                           payload.dump(),
                           "application/json");
    if (res && (res->status == 200 || res->status == 204)) {
        std::cout << "[" << now() << "] Message sent: " << payload["content"] << "\n";
    } else {
        std::cout << "[" << now() << "] Failed to send message: " << (res ? res->status : -1) << "\n";
    }
}

void send_embed(httplib::Client& client, const std::string& channel_id, const json& embed) {
    json payload = {{"embeds", {embed}}};
    send_message(client, channel_id, payload);
}

void create_channel(httplib::Client& client, const std::string& guild_id, const std::string& name, const std::string& category_id) {
    json payload = {
        {"name", name},
        {"type", 0}, // GUILD_TEXT
        {"parent_id", category_id}
    };
    auto res = client.Post(("/guilds/" + guild_id + "/channels").c_str(),
                           "Bot " + BOT_TOKEN,
                           payload.dump(),
                           "application/json");
    if (res && res->status == 200) {
        std::cout << "[" << now() << "] Created channel: " << name << "\n";
    } else {
        std::cout << "[" << now() << "] Failed to create channel: " << (res ? res->status : -1) << "\n";
    }
}

void delete_channel(httplib::Client& client, const std::string& channel_id) {
    auto res = client.Delete(("/channels/" + channel_id).c_str(),
                             "Bot " + BOT_TOKEN,
                             "application/json");
    if (res && (res->status == 200 || res->status == 204)) {
        std::cout << "[" << now() << "] Deleted channel: " << channel_id << "\n";
    } else {
        std::cout << "[" << now() << "] Failed to delete channel: " << (res ? res->status : -1) << "\n";
    }
}

void on_ready(httplib::Client& client) {
    std::cout << "[" << now() << "] Bot online\n";
}

void handle_message(httplib::Client& client, const json& message) {
    if (message["author"]["bot"].get<bool>()) return;

    if (message["channel_id"].get<std::string>().find("dm") != std::string::npos) {
        std::string user_id = message["author"]["id"].get<std::string>();
        std::string channel_name = "modmail-" + user_id;

        // Check for category
        auto category_res = client.Get(("/guilds/" + GUILD_ID + "/channels").c_str(),
                                      "Bot " + BOT_TOKEN);
        std::string category_id;
        if (category_res && category_res->status == 200) {
            json channels = json::parse(category_res->body);
            for (const auto& channel : channels) {
                if (channel["type"] == 4 && channel["name"] == MODMAIL_CATEGORY_NAME) {
                    category_id = channel["id"].get<std::string>();
                    break;
                }
            }
        }

        if (category_id.empty()) {
            json payload = {{"name", MODMAIL_CATEGORY_NAME}, {"type", 4}};
            auto res = client.Post(("/guilds/" + GUILD_ID + "/channels").c_str(),
                                   "Bot " + BOT_TOKEN,
                                   payload.dump(),
                                   "application/json");
            if (res && res->status == 200) {
                category_id = json::parse(res->body)["id"].get<std::string>();
            }
        }

        // Check for existing channel
        std::string channel_id;
        if (!category_id.empty()) {
            auto res = client.Get(("/guilds/" + GUILD_ID + "/channels").c_str(),
                                  "Bot " + BOT_TOKEN);
            if (res && res->status == 200) {
                json channels = json::parse(res->body);
                for (const auto& channel : channels) {
                    if (channel["name"] == channel_name) {
                        channel_id = channel["id"].get<std::string>();
                        break;
                    }
                }
            }
        }

        if (channel_id.empty()) {
            create_channel(client, GUILD_ID, channel_name, category_id);
            // Fetch new channel ID (simplified; in practice, use the response from create_channel)
            auto res = client.Get(("/guilds/" + GUILD_ID + "/channels").c_str(),
                                  "Bot " + BOT_TOKEN);
            if (res && res->status == 200) {
                json channels = json::parse(res->body);
                for (const auto& channel : channels) {
                    if (channel["name"] == channel_name) {
                        channel_id = channel["id"].get<std::string>();
                        break;
                    }
                }
            }

            json message_payload = {{"content", "📬 New ModMail started by <@" + user_id + ">"}};
            send_message(client, channel_id, message_payload);
            json user_message = {{"content", "**" + message["author"]["username"].get<std::string>() + ":** " + message["content"].get<std::string>()}};
            send_message(client, channel_id, user_message);

            json user_response = {{"content", "✅ Your message has been sent to the support team!"}};
            send_message(client, message["channel_id"].get<std::string>(), user_response);

            ticket_start_times[channel_id] = now();

            json log_embed = {
                {"title", "📉 New Ticket"},
                {"description", message["author"]["username"].get<std::string>() + " | `" + user_id + "` • " + now()},
                {"color", 0x5865F2}
            };
            send_embed(client, MODMAIL_LOG_CHANNEL, log_embed);
        } else {
            json user_message = {{"content", "**" + message["author"]["username"].get<std::string>() + ":** " + message["content"].get<std::string>()}};
            send_message(client, channel_id, user_message);
        }
    } else if (message["guild_id"] == GUILD_ID) {
        auto res = client.Get(("/channels/" + message["channel_id"].get<std::string>()).c_str(),
                              "Bot " + BOT_TOKEN);
        if (res && res->status == 200) {
            json channel = json::parse(res->body);
            if (channel["parent_id"].is_string()) {
                auto category_res = client.Get(("/channels/" + channel["parent_id"].get<std::string>()).c_str(),
                                               "Bot " + BOT_TOKEN);
                if (category_res && category_res->status == 200) {
                    json category = json::parse(category_res->body);
                    if (category["name"] == MODMAIL_CATEGORY_NAME) {
                        std::string user_id = channel["name"].get<std::string>().replace(0, 8, "");
                        json user_message = {{"content", message["content"].get<std::string>()}};
                        send_message(client, user_id, user_message);
                    }
                }
            }
        }
    }
}

void handle_close_command(httplib::Client& client, const json& message, const std::string& reason) {
    auto res = client.Get(("/channels/" + message["channel_id"].get<std::string>()).c_str(),
                          "Bot " + BOT_TOKEN);
    if (res && res->status == 200) {
        json channel = json::parse(res->body);
        if (!channel["parent_id"].is_string() || channel["name"] != ("modmail-" + message["author"]["id"].get<std::string>())) {
            json payload = {{"content", "❌ This command can only be used in a modmail thread."}};
            send_message(client, message["channel_id"].get<std::string>(), payload);
            return;
        }
    }

    if (reason.empty()) {
        json payload = {{"content", "❌ You must provide a reason. Usage: `!close <reason>`"}};
        send_message(client, message["channel_id"].get<std::string>(), payload);
        return;
    }

    std::string user_id = message["channel_id"].get<std::string>().replace(0, 8, "");
    std::string start_time = ticket_start_times[message["channel_id"].get<std::string>()];
    auto now_time = std::time(nullptr);
    auto start_time_t = std::mktime(std::localtime(&now_time));
    std::string formatted_duration = start_time.empty() ? "Unknown" : std::to_string((now_time - start_time_t) / 3600) + "h " + std::to_string((now_time - start_time_t % 3600) / 60) + "m " + std::to_string(now_time - start_time_t % 60) + "s";

    json embed = {
        {"title", "📫 ModMail Closed"},
        {"color", 0xFF0000},
        {"fields", {
            {{"name", "Closed By"}, {"value", message["author"]["username"].get<std::string>() + " (`" + message["author"]["id"].get<std::string>() + "`)"}, {"inline", false}},
            {{"name", "Time Open"}, {"value", formatted_duration}, {"inline", false}},
            {{"name", "Reason"}, {"value", reason}, {"inline", false}}
        }},
        {"timestamp", now()}
    };
    send_embed(client, user_id, embed);

    json log_embed = {
        {"title", "🗑️ Ticket Closed"},
        {"description", "Thread: <#" + message["channel_id"].get<std::string>() + "> was closed."},
        {"color", 0xFF0000},
        {"fields", embed["fields"]}
    };
    send_embed(client, MODMAIL_LOG_CHANNEL, log_embed);

    json response = {{"content", "✅ Ticket closed."}};
    send_message(client, message["channel_id"].get<std::string>(), response);

    delete_channel(client, message["channel_id"].get<std::string>());
}

int main() {
    httplib::Client client("https://discord.com/api/v10");
    client.set_default_headers({{"Authorization", "Bot " + BOT_TOKEN}});

    on_ready(client);

    // Simulate message handling (in practice, use WebSocket or DPP)
    json sample_message = {
        {"author", {{"id", "user_id"}, {"username", "TestUser"}, {"bot", false}}},
        {"channel_id", "channel_id"},
        {"guild_id", GUILD_ID},
        {"content", "Hello!"}
    };
    handle_message(client, sample_message);

    json close_command = {
        {"author", {{"id", "user_id"}, {"username", "TestUser"}, {"bot", false}}},
        {"channel_id", "modmail-user_id"},
        {"guild_id", GUILD_ID},
        {"content", "!close Test reason"}
    };
    handle_close_command(client, close_command, "Test reason");

    while (true) {
        std::this_thread::sleep_for(std::chrono::seconds(10));
        std::cout << "[" << now() << "] Bot running...\n";
    }

    return 0;
}

File: py/version_1.0.0.py
import discord
from discord.ext import commands
from discord.ui import View, Button, Modal, TextInput
import datetime

GUILD_ID = ts server_id  # Replace with your server ID
MODMAIL_CATEGORY_NAME = "ModMail" # create category named ModMail or change it urself
bot_token = "paste token" # Replace with your token

intents = discord.Intents.default()
intents.messages = True
intents.guilds = True
intents.dm_messages = True
intents.message_content = True
intents.members = True

bot = commands.Bot(command_prefix="!", intents=intents)
ticket_start_times = {}  # Store when a modmail channel is created

class CloseReasonModal(Modal, title="Close ModMail Ticket"):
    reason = TextInput(label="Reason for closing", style=discord.TextStyle.paragraph)

    def __init__(self, user, staff, channel):
        super().__init__()
        self.user = user
        self.staff = staff
        self.channel = channel

    async def on_submit(self, interaction: discord.Interaction):
        now = datetime.datetime.utcnow()
        open_time = ticket_start_times.get(self.channel.id)
        duration = now - open_time if open_time else None
        formatted_duration = str(duration).split('.')[0] if duration else "Unknown"

        embed = discord.Embed(title="\ud83d\udceb ModMail Closed", color=discord.Color.red())
        embed.add_field(name="Closed By", value=f"{self.staff} (`{self.staff.id}`)", inline=False)
        embed.add_field(name="Time Open", value=formatted_duration, inline=False)
        embed.add_field(name="Reason", value=self.reason.value, inline=False)
        embed.timestamp = now

        try:
            await self.user.send(embed=embed)
        except:
            await interaction.channel.send("\u26a0\ufe0f Could not notify the user.")

        log_channel = discord.utils.get(interaction.guild.text_channels, name="modmail-log")
        if log_channel:
            log_embed = embed.copy()
            log_embed.title = "\ud83d\uddd1\ufe0f Ticket Closed"
            log_embed.description = f"Thread: {self.channel.mention} was closed."
            await log_channel.send(embed=log_embed)

        await interaction.response.send_message("\u2705 Ticket closed.", ephemeral=True)
        await self.channel.delete()

class CloseTicketView(View):
    def __init__(self, user, staff, channel):
        super().__init__(timeout=None)
        self.user = user
        self.staff = staff
        self.channel = channel

    @discord.ui.button(label="Close", style=discord.ButtonStyle.danger, emoji="\ud83d\uddd1\ufe0f", custom_id="close_button")
    async def close(self, interaction: discord.Interaction, button: discord.ui.Button):
        if not interaction.user.guild_permissions.manage_channels:
            await interaction.response.send_message("You do not have permission to close this.", ephemeral=True)
            return
        modal = CloseReasonModal(user=self.user, staff=interaction.user, channel=self.channel)
        await interaction.response.send_modal(modal)

@bot.event
async def on_ready():
    print(f"\u2705 Bot online as {bot.user}")

@bot.event
async def on_message(message):
    if message.author.bot:
        return

    if isinstance(message.channel, discord.DMChannel):
        guild = bot.get_guild(GUILD_ID)
        if guild is None:
            return

        category = discord.utils.get(guild.categories, name=MODMAIL_CATEGORY_NAME)
        if category is None:
            category = await guild.create_category(MODMAIL_CATEGORY_NAME)

        channel_name = f"modmail-{message.author.id}"
        existing_channel = discord.utils.get(guild.text_channels, name=channel_name)

        if not existing_channel:
            new_channel = await guild.create_text_channel(channel_name, category=category)
            await new_channel.send(
                f"\ud83d\udcec New ModMail started by {message.author.mention}",
                view=CloseTicketView(user=message.author, staff=bot.user, channel=new_channel)
            )
            await new_channel.send(f"**{message.author}:** {message.content}")
            for a in message.attachments:
                await new_channel.send(file=await a.to_file())
            await message.author.send("\u2705 Your message has been sent to the support team!")
            ticket_start_times[new_channel.id] = datetime.datetime.utcnow()

            log_channel = discord.utils.get(guild.text_channels, name="modmail-log")
            if log_channel:
                embed = discord.Embed(title="\ud83d\udcc9 New Ticket", color=discord.Color.blurple())
                timestamp = discord.utils.format_dt(datetime.datetime.utcnow(), style='f')
                embed.description = f"_{message.author} | `{message.author.id}` \u2022 {timestamp}"
                await log_channel.send(embed=embed)

        else:
            await existing_channel.send(f"**{message.author}:** {message.content}")
            for a in message.attachments:
                await existing_channel.send(file=await a.to_file())

    elif message.guild and message.channel.category and message.channel.category.name == MODMAIL_CATEGORY_NAME:
        try:
            user_id = int(message.channel.name.replace("modmail-", ""))
            user = await bot.fetch_user(user_id)
            if message.content:
                await user.send(message.content)
            for a in message.attachments:
                await user.send(file=await a.to_file())
        except:
            await message.channel.send("\u274c Could not send message to user.")

    await bot.process_commands(message)

@bot.command()
@commands.has_permissions(manage_channels=True)
async def close(ctx, *, reason: str = None):
    if not ctx.channel.category or ctx.channel.category.name != MODMAIL_CATEGORY_NAME:
        await ctx.send("\u274c This command can only be used in a modmail thread.")
        return

    if not reason:
        await ctx.send("\u274c You must provide a reason. Usage: `!close <reason>`")
        return

    try:
        await ctx.message.delete()
    except:
        pass

    try:
        user_id = int(ctx.channel.name.replace("modmail-", ""))
        user = await bot.fetch_user(user_id)

        start_time = ticket_start_times.get(ctx.channel.id)
        now = datetime.datetime.utcnow()
        duration = now - start_time if start_time else None
        formatted_duration = str(duration).split('.')[0] if duration else "Unknown"

        embed = discord.Embed(title="\ud83d\udceb ModMail Closed", color=discord.Color.red())
        embed.add_field(name="Closed By", value=f"{ctx.author} (`{ctx.author.id}`)", inline=False)
        embed.add_field(name="Time Open", value=formatted_duration, inline=False)
        embed.add_field(name="Reason", value=reason, inline=False)
        embed.timestamp = now

        await user.send(embed=embed)

        log_channel = discord.utils.get(ctx.guild.text_channels, name="modmail-log")
        if log_channel:
            log_embed = embed.copy()
            log_embed.title = "\ud83d\uddd1\ufe0f Ticket Closed"
            log_embed.description = f"Thread: {ctx.channel.mention} was closed."
            await log_channel.send(embed=log_embed)

    except:
        await ctx.send("\u26a0\ufe0f Could not DM the user.")

    await ctx.send("\u2705 Ticket closed.")
    await ctx.channel.delete()

bot.run(bot_token)
